compiler/noirc_frontend/src/node_interner.rs

```diff
@@ -916,3 +942,26 @@ fn get_type_method_key(typ: &Type) -> Option<TypeMethodKey> {
         | Type::FmtString(_, _) => None,
     }
 }
+
+pub fn allow_trait_impl_for_type(typ: &Type) -> bool {
+    match &typ {
+        Type::FieldElement
+        | Type::Array(..)
+        | Type::Integer(..)
+        | Type::Bool
+        | Type::Struct(..)
+        | Type::Tuple(..)
+        | Type::String(..) => true,
+
+        Type::FmtString(..)
+        | Type::Unit
+        | Type::TypeVariable(..)
+        | Type::NamedGeneric(..)
+        | Type::Function(..)
+        | Type::MutableReference(..)
```

Comment:
We should still allow primitive impls on fmtstrings, unit, and functions. Mutable references could be useful too, but there may be some conflicts with auto-dereferencing that'd make us want to hold off.

Reply:
I extended the implementation to include  `Type::FmtString(..) +  Type::Unit`, but I can't figure out how to write test for the rest. Can you suggest something ?

Reply:
Which types do you mean by "for the rest?" It looks like you have all the currently passing cases covered. Once you add functions you can write an impl such as: `impl Foo for fn(u32) -> u32 { ... }`.

For the failing tests, we don't need these for all types since most of the ones that are prohibited are internal types (like `NotConstant`, `TypeVariable`, `Forall`, and `Error`) that aren't possible for users to write anyway. The only two that may be useful to prohibit impls for are mutable reference (debatable, since you can write impls on references in rust), and constants. An impl for the later would look like `impl Foo for 3 { ... }` which in most cases the parser already syntactically disallows.

Reply:
Fixed in  [ecabf0](https://github.com/noir-lang/noir/commit/ecabf02e4d27536b9a83a9f4271e4c7f35aaa9fe) Added & tested support for Function & MutableReference